pkg/apis/stork/v1alpha1/types.go

```diff
@@ -212,3 +217,193 @@ type MigrationList struct {
+	// This will be ignored if Devices is not empty.
+	UseAll bool `json:"useAll"`
+	// UseAllWithPartitions use all available unformatted devices
+	// including partitions. This will be ignored if Devices is not empty.
+	UseAllWithPartitions bool `json:"useAllWithPartitions"`
+	// Devices list of devices to be used by storage driver
+	Devices []string `json:"devices"`
+	// StorageMdDevices metadata device for storage pools
+	StorageMdDevices []string `json:"storageMetadataDevices"`
+	// JournalDevice device for journaling
+	JournalDevice string `json:"journalDevice"`
+	// SystemMdDevice device that will be used to store system metadata
+	SystemMdDevice string `json:"systemMetadataDevice"`
+	// DataStorageType type of data storage used
+	DataStorageType string `json:"dataStorageType"`
+	// RaidLevel raid level for the storage pool
+	RaidLevel string `json:"raidLevel"`
+}
+
+// CloudStorageSpec details of storage in cloud environment
+type CloudStorageSpec struct {
+	// DeviceSpecs list of storage device specs. A cloud storage device will
+	// be created for every spec in the DeviceSpecs list. Currently,
+	// CloudStorageSpec is only at the cluster level, so the below specs
+	// be applied to all storage nodes in the cluster.
+	DeviceSpecs []string `json:"deviceSpecs"`
+	// JournalDeviceSpec spec for the journal device
+	JournalDeviceSpec string `json:"journalDeviceSpec"`
+	// SystemMdDeviceSpec spec for the metadata device
+	SystemMdDeviceSpec string `json:"systemMetadataDeviceSpec"`
+	// MaxStorageNodes maximum nodes that will have storage in the cluster
+	MaxStorageNodes uint32 `json:"maxStorageNodes"`
+	// MaxStorageNodesPerZone maximum nodes in every zone that will have
+	// storage in the cluster
+	MaxStorageNodesPerZone uint32 `json:"maxStorageNodesPerZone"`
+}
+
+// Geography is topology information for a node
+type Geography struct {
+	// Region region in which the node is placed
+	Region string `json:"region"`
+	// Zone zone in which the node is placed
+	Zone string `json:"zone"`
+	// Rack rack on which the node is placed
+	Rack string `json:"rack"`
+}
+
+// StorageClusterStatus is the status of a storage cluster
+type StorageClusterStatus struct {
+	// ClusterName name of the storage cluster
+	ClusterName string `json:"clusterName"`
+	// ClusterUUID uuid for the storage cluster
+	ClusterUUID string `json:"clusterUuid"`
+	// CreatedAt timestamp at which the storage cluster was created
+	CreatedAt meta.Time `json:"createdAt"`
+	// Status status of the storage cluster
+	Status string `json:"status"`
+	// NodeStatuses list of statuses for all the nodes in the storage cluster
+	NodeStatuses []NodeStatus `json:nodes`
+}
+
+// NodeStatus status of the storage cluster node
+type NodeStatus struct {
+	// NodeName name of the node
+	NodeName string `json:"nodeName"`
+	// NodeUUID uuid of the node
+	NodeUUID string `json:"nodeUuid"`
+	// Status status of the node
+	Status string `json:"status"`
```

Comment:
Status needs to be an enum. Have a StatusMessage for arbitrary string status descriptions.

Reply:
For node conditions, I prefer how k8s does it.

https://github.com/portworx/porx/blob/master/vendor/k8s.io/api/core/v1/types.go#L3950

A node status can have a list of conditions, each with it's own state and msg. Hence a PX node can express multiple conditions (e.g Storage down, Control plane up, Maintenance mode etc.) at the same time.

A single status will be limiting.